extensions/sftp/processors/ListSFTP.cpp

```diff
@@ -0,0 +1,1320 @@
+        logger_->log_error("Tracking Entities state json file \"%s\" timestamp missing or malformatted for entity \"%s\"",
+            state_json_state_file.c_str(),
+            already_listed_entity.name.GetString());
+        continue;
+      }
+      uint64_t timestamp = it->value.GetUint64();
+      it = already_listed_entity.value.FindMember("size");
+      if (it == already_listed_entity.value.MemberEnd() || !it->value.IsUint64()) {
+        logger_->log_error("Tracking Entities state json file \"%s\" size missing or malformatted for entity \"%s\"",
+                           state_json_state_file.c_str(),
+                           already_listed_entity.name.GetString());
+        continue;
+      }
+      uint64_t size = it->value.GetUint64();
+      new_already_listed_entities.emplace(std::piecewise_construct,
+                                          std::forward_as_tuple(already_listed_entity.name.GetString()),
+                                          std::forward_as_tuple(timestamp, size));
+    }
+    already_listed_entities_ = std::move(new_already_listed_entities);
+  } catch (std::exception& e) {
+    logger_->log_error("Exception while parsing Tracking Entities state json file \"%s\": %s", state_json_state_file.c_str(), e.what());
+    return false;
+  }
+
+  return true;
+}
+
+void ListSFTP::listByTrackingEntities(
+    const std::shared_ptr<core::ProcessContext>& context,
+    const std::shared_ptr<core::ProcessSession>& session,
+    const std::string& hostname,
+    uint16_t port,
+    const std::string& username,
+    const std::string& remote_path,
+    uint64_t entity_tracking_time_window,
+    std::vector<Child>&& files) {
+  /* Load state from cache file if needed */
+  if (!already_loaded_from_cache_ && !tracking_entities_state_filename_.empty()) {
+    if (updateFromTrackingEntitiesCache(hostname, username, remote_path)) {
+      logger_->log_debug("Successfully loaded Tracking Entities state file \"%s\"", tracking_entities_state_filename_.c_str());
+      initial_listing_complete_ = true;
+    } else {
+      logger_->log_debug("Failed to load Tracking Entities state file \"%s\"", tracking_entities_state_filename_.c_str());
+    }
+    already_loaded_from_cache_ = true;
+  }
+
+  time_t now = time(nullptr);
+  uint64_t min_timestamp_to_list = (!initial_listing_complete_ && entity_tracking_initial_listing_target_ == ENTITY_TRACKING_INITIAL_LISTING_TARGET_ALL_AVAILABLE)
+      ? 0U : (now * 1000 - entity_tracking_time_window);
+
+  /* Skip files not in the tracking window */
+  for (auto it = files.begin(); it != files.end(); ) {
+    if (it->attrs.mtime * 1000 < min_timestamp_to_list) {
+      logger_->log_trace("Skipping \"%s\" because it has an older timestamp than the minimum timestamp to list: %lu < %lu",
+          it->getPath(), it->attrs.mtime * 1000, min_timestamp_to_list);
+      it = files.erase(it);
+    } else {
+      ++it;
+    }
+  }
+
+  if (files.size() == 0U) {
```

Comment:
empty? :)

Reply:
Yeah, I have temporarily forgotten about that, thanks.